### PR TITLE
Refactored generator_v3

### PR DESCRIPTION
**Pull request by Joshimuz**
Removed old comments, added new comments
Removed unused variables
Better creation of Indexes array
Added tagCount list which keeps track of tags on the sheet and their amounts instead of recalculating the list/amount by checking every goal on the sheet every time a goal is generated!
Added antisynergy, reactant and catalyst arrays, to keep track of what antisyergys etc are on the sheet instead of checking every goal on the sheet every time a goal is generated!
Refactored check for if the goal is already on the sheet.
Refactored check for same line tag collisions.
Renamed cont to retry, moved it outside the do while loop (how did this even work anyway?).
Removed constant refrences to cont when failing a goal check and instead just set retry to false ONCE at the end of the do while loop is all the checks are passed.
Named do while loop GoalGen and continue that specifically when failing a check.

This addresses #98 AND THEN SOME

### CODE REVIEW

File: generator_v3.js [hunk 1]
```diff
@@ -107,28 +106,36 @@ var generator_v3 = function(layout, difficulty, bingoList)
...
+	reactants = [],
+	catalysts = [];
+
+	// Try to generate 25 goals to populate the sheet
 	for (var i=0; i<=24; i++)
 	{
+		// Keep track of how many times we've tried to generate a goal
 		var failSafe = 0;
 
+		// cont keeps track of if current goalCandidate is good and should be added or not
```
Comment (rybak) [inline generator_v3.js:125]: Variable has been renamed to `retry`, so the comment should be updated as well.

File: generator_v3.js [hunk 2]
```diff
@@ -140,90 +147,93 @@ var generator_v3 = function(layout, difficulty, bingoList)
...
+						// If both goals have the same tag
+						if (currentSheet[indexes[z]].tags.some(r=> r.line == false && goalCandidate.tags.some(s=> r.name === s.name)))
+						{
+							console.log("Cannot be on same line: " + goalCandidate.name + " and " + currentSheet[indexes[z]].name);
+							continue GoalGen;
+						}
 					}
 				}
-			}
+			}
```
Comment (rybak) [inline generator_v3.js:236]: Unnecessary whitespace added at the end of this line.

Comment (Joshimuz) [reply generator_v3.js]: You do love your whitespace

**Comment by rybak** [comment]
I can't do a full review right now. Will continue tomorrow, probably.

**Comment by Joshimuz** [comment]
I'll just assume it's fine :)

File: generator_v3.js [hunk 3]
```diff
@@ -140,87 +147,90 @@ var generator_v3 = function(layout, difficulty, bingoList)
...
+					// If it is, get a new goal
+					console.log("catalyst for: " + goalCandidate.name);
+					continue GoalGen;
 				}
 			}
 
-			// If the tag count is higher than the max that tag should have, don't continue
-			if (cont && goalCandidate.tags != null)
+			// If the goal candidate contains tags that cannot be on the same line as other goals with that tag
+			if (goalCandidate.tags.some(r=> r.line == false))
```
Comment (rybak) [inline generator_v3.js:220]: This `if`, which checks `.line` of the tags is outside of the `if (goalCandidate.tags != null)` above, which is suspicious.

Comment (rybak) [reply generator_v3.js:220]: Either both of these `if`s should check `.tags` for nullity or we should assume that `.tags` is not `null` for all goals.

Comment (rybak) [reply generator_v3.js:220]: Oh, there actually exists such an `if` for `goalCandidate`. It is just below, under comment "// If both goals have tags AND are on the same line". Except that it doesn't make sense to check `goalCandidate.tags` for `null` when `.some(..)` has already been called on it.

File: generator_v3.js [hunk 4]
```diff
@@ -140,87 +147,90 @@ var generator_v3 = function(layout, difficulty, bingoList)
...
-				// Check if the goal generated has any anti synergy with anything already on the sheet
-				else if (typeof currentSheet[indexes[z]].antisynergy !== 'undefined' && typeof goalCandidate.antisynergy !== 'undefined'
-					&& currentSheet[indexes[z]].antisynergy.some(r=> goalCandidate.antisynergy.includes(r)))
+			}
+
+			// Check if the goal (and the goals on the sheet) has any antisynergies
+			if (typeof goalCandidate.antisynergy !== 'undefined')
+			{
+				// If it does, check to see if it's already on the sheet
+				if (antisynergys.some(r=> goalCandidate.antisynergy.includes(r)))
```
Comment (rybak) [inline generator_v3.js:190]: Here, `goalCandidate.antisynergy` is a small list (every antisynergy in `goals_v4.js` has only one element), while `antisynergys` is a big list (constantly growing from zero up to 25). So it would make more sense to compare them the other way around.

Same argument applies to `goalCandidate.catalyst` and `goalCandidate.reactant`

Comment (rybak) [reply generator_v3.js:190]: This would make even more sense, if `antisynergys`, `reactants`, and `catalysts` were a [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).

Comment (rybak) [reply generator_v3.js:190]: Opened #100.

File: generator_v3.js (hunk 3 above)
Comment (rybak) [reply generator_v3.js:220]: Fixed in da9b544.
